### PR TITLE
fix:[PL-41308]: Fix to fetch secrets in batch when the number of secrets are more than the default unlimited

### PR DESCRIPTION
**Pull request by nishantsaini20**
### Issue:
The vaule of `DEFAULT_UNLIMITED` currently is **1000**. Due to this not more than **1000** secrets are fetched in list secret API. This has more impact where scope filter is applied. E.g. fetching secrets in application scope. The scope filtered is applied after fetching secrets from DB. Since `DEFAULT_UNLIMITED` is **1000** it the scoping filter is applied on on first **1000** secrets and rest get ignored and never get returned in the response.
### Solution:
Keep fetching and applying filter on the secrets in batches and accumulate the filtered secrets till the required number of secrets for a given page are not accumulated or no more secrets are returned by the query.

### Testing:
On local env changed the `DEFAULT_UNLIMITED` to **5** and validated the following:
1. With `limit` **3** and `offsets` **0, 3, 6** should return expected number of secret
2. Without passing `limit` and `offset` total of **5** secrets (i.e. equal to `DEFAULT_UNLIMITED`) should be returned when total secrets of a given type are more than **5**.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53082)

### CODE REVIEW

**Comment by nishantsaini20** [comment]
trigger uts

File: 890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java [hunk 1]
```diff
@@ -791,28 +791,37 @@ public PageResponse<EncryptedData> listSecrets(
 
     try {
       PageRequest<EncryptedData> copiedPageRequest = pageRequest.copy();
-      int offset = copiedPageRequest.getStart();
+      int offset = 0;
       int limit = copiedPageRequest.getPageSize();
+      int total;
+      int numberOfRecordsRequiredToFetch = pageRequest.getStart() + pageRequest.getPageSize();
```
Comment (meenakshi-harness) [inline 890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java:797]: Didn't get this. we can discuss on call

File: 890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java [hunk 2]
```diff
@@ -791,28 +791,37 @@ public PageResponse<EncryptedData> listSecrets(
...
+      List<EncryptedData> response;
 
-      copiedPageRequest.setOffset("0");
       copiedPageRequest.setLimit(String.valueOf(PageRequest.DEFAULT_UNLIMITED));
       if (limit > PageRequest.DEFAULT_UNLIMITED) {
         copiedPageRequest.setLimit(String.valueOf(limit));
```
Comment (meenakshi-harness) [inline 890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java:802]: Shouldn't we setLimit with `limit` passed in PageRequest alway except when `limit` is null set it to `DEFAULT_UNLIMITED` ?

File: 890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java [hunk 3]
```diff
@@ -791,28 +791,37 @@ public PageResponse<EncryptedData> listSecrets(
...
-      List<EncryptedData> encryptedDataList = batchPageResponse.getResponse();
-      filterSecreteDataBasedOnUsageRestrictions(accountId, appId, envId, encryptedDataList, filteredEncryptedDataList);
-      List<EncryptedData> response;
-      if (isNotEmpty(filteredEncryptedDataList) && filteredEncryptedDataList.size() > offset) {
-        int endIdx = Math.min(offset + limit, filteredEncryptedDataList.size());
-        response = filteredEncryptedDataList.subList(offset, endIdx);
-      } else {
-        response = Collections.emptyList();
+      while (true) {
+        copiedPageRequest.setOffset(String.valueOf(offset));
```
Comment (meenakshi-harness) [inline 890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java:805]: why are we not taking offset from pagerequest? Lets discuss the changes on call.

**Review by meenakshi-harness — CHANGES_REQUESTED** [review]

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53082)* status: 2 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java` line 797 at r1](https://reviewable.io/reviews/harness/harness-core/53082#gh-1333225857:-Newh_e539wLvab-s5xH:b-xzk65y) ([raw file](https://github.com/harness/harness-core/blob/769708bfee19de68e9e38116ff606b691ccccb2b/890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java#L797)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

Didn't get this. we can discuss on call
</blockquote></details>

As disscussed, this is required to stop fetching more secrets from db once we have the sufficient number pulled to serve the requested page.
___
*[`890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java` line 802 at r1](https://reviewable.io/reviews/harness/harness-core/53082#gh-1333227884:-NewhomJ1RyruxJXtFCG:b-emqnuc) ([raw file](https://github.com/harness/harness-core/blob/769708bfee19de68e9e38116ff606b691ccccb2b/890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java#L802)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

Shouldn't we setLimit with `limit` passed in PageRequest alway except when `limit` is null set it to `DEFAULT_UNLIMITED` ?
</blockquote></details>

This change was done as part of a different ticket where a higher limit is allow based on a feature flag.
___
*[`890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java` line 805 at r1](https://reviewable.io/reviews/harness/harness-core/53082#gh-1333229513:-Newi6zB6kpVAhl8-_4S:b-mm8dh6) ([raw file](https://github.com/harness/harness-core/blob/769708bfee19de68e9e38116ff606b691ccccb2b/890-sm-core/src/main/java/io/harness/secrets/SecretServiceImpl.java#L805)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

why are we not taking offset from pagerequest? Lets discuss the changes on call.
</blockquote></details>

As discussed, that can result in returning secret which we returned for previous page because filtering is not part of db query but done via business logic after pulling secrets from db.

**Comment by nishantsaini20** [comment]
trigger uts

**Comment by nishantsaini20** [comment]
trigger uts

**Review by meenakshi-harness — APPROVED** [review]

**Pull request merged**